src/main/java/org/icatproject/core/manager/GateKeeper.java

```diff
@@ -164,35 +164,102 @@ private void exit() {
 		return publicTables;
 	}
 
-	public List<EntityBaseBean> getReadable(String userId, List<EntityBaseBean> beans, EntityManager manager) {
-
-		if (beans.size() == 0) {
-			return beans;
-		}
-
-		EntityBaseBean object = beans.get(0);
-
-		Class<? extends EntityBaseBean> objectClass = object.getClass();
-		String simpleName = objectClass.getSimpleName();
+	/**
+	 * @param userId     The user making the READ request
+	 * @param simpleName The name of the requested entity type
+	 * @param manager
+	 * @return Returns a list of restrictions that apply to the requested entity
+	 *         type. If there are no restrictions, then returns null.
+	 */
+	private List<String> getRestrictions(String userId, String simpleName, EntityManager manager) {
 		if (rootUserNames.contains(userId)) {
 			logger.info("\"Root\" user " + userId + " is allowed READ to " + simpleName);
-			return beans;
+			return null;
 		}
 
 		TypedQuery<String> query = manager.createNamedQuery(Rule.INCLUDE_QUERY, String.class)
 				.setParameter("member", userId).setParameter("bean", simpleName);
 
 		List<String> restrictions = query.getResultList();
 		logger.debug("Got " + restrictions.size() + " authz queries for READ by " + userId + " to a "
-				+ objectClass.getSimpleName());
+				+ simpleName);
 
 		for (String restriction : restrictions) {
 			logger.debug("Query: " + restriction);
 			if (restriction == null) {
 				logger.info("Null restriction => READ permitted to " + simpleName);
-				return beans;
+				return null;
+			}
+		}
+
+		return restrictions;
+	}
+
+	/**
+	 * Returns a sub list of the passed entities that the user has READ access to.
+	 * 
+	 * @param userId  The user making the READ request
+	 * @param beans   The entities the user wants to READ
+	 * @param manager
+	 * @return A list of entities the user has read access to
+	 */
+	public List<EntityBaseBean> getReadable(String userId, List<EntityBaseBean> beans, EntityManager manager) {
+
+		if (beans.size() == 0) {
+			return beans;
+		}
+		EntityBaseBean object = beans.get(0);
+		Class<? extends EntityBaseBean> objectClass = object.getClass();
+		String simpleName = objectClass.getSimpleName();
+
+		List<String> restrictions = getRestrictions(userId, simpleName, manager);
+		if (restrictions == null) {
+			return beans;
+		}
+
+		Set<Long> readableIds = getReadableIds(userId, beans, restrictions, manager);
+
+		List<EntityBaseBean> results = new ArrayList<>();
+		for (EntityBaseBean bean : beans) {
+			if (readableIds.contains(bean.getId())) {
+				results.add(bean);
 			}
 		}
+		return results;
+	}
+
+	/**
+	 * @param userId     The user making the READ request
```

Comment:
Add a method description as well, perhaps using some of the description for the return type.